nemo_aligner/utils/distributed.py

```diff
@@ -191,12 +193,114 @@ def _compute_distributed_log_softmax(vocab_parallel_logits):
+
+    # Gather top-k values and indices from all GPUs
+    torch.distributed.gather(local_topk_values, gathered_values if rank == 0 else None, dst=0)
+    torch.distributed.gather(local_topk_indices, gathered_indices if rank == 0 else None, dst=0)
+
+    if rank == 0:  # only rank 0 will do the computation and scatter the outcome
+        # Concatenate the gathered values and indices along a new dimension
+        all_values = torch.cat(gathered_values, dim=-1)  # [B,S,world_size*k]
+        all_indices = torch.cat(gathered_indices, dim=-1)
+
+        # Perform a global top-k operation to find the global top-k values and indices
+        global_topk_values, topk_indices = torch.topk(all_values, k=k, dim=-1)  # [B,S,k]
+        global_topk_indices = torch.gather(all_indices, -1, topk_indices)
+
+        # perform top_p
+        if p > 0.0:
+            # todo: perform top_p and save in global_top_k_p_indices and spread to all ranks
+            # meanwhile...
+            sorted_logits, sorted_indices = torch.sort(global_topk_values, descending=True, dim=-1)  # [B,S,k] for each
+            cumulative_probs = torch.cumsum(F.softmax(sorted_logits, dim=-1), dim=-1)  # [B,S,k]
+            global_top_k_p_indices = torch.gather(global_topk_indices, -1, sorted_indices)
+            sorted_indices_to_remove = cumulative_probs > p
+            global_top_k_p_indices = torch.where(
+                sorted_indices_to_remove, torch.tensor(-1, dtype=torch.long), global_top_k_p_indices
+            )  # the top_p are kept, the rest are set to -1 (so will be filtered in the following lines)
+    else:
+        global_top_k_p_indices = torch.empty_like(
+            local_topk_indices, dtype=local_topk_indices.dtype, device=local_topk_indices.device
+        )  # [B,S,k]
+
+    torch.distributed.broadcast(global_top_k_p_indices, src=0)
+
+    # generate a mask according to the rank
+    # filter indices within the current rank's segment
+    mask_top_k_p_indices = (global_top_k_p_indices >= vocab_start_index) & (
+        global_top_k_p_indices < vocab_end_index
+    )  # [B,S,k] where only indices that are within the scope of current rank are True
+
+    # adjust indices to local index space
+    local_top_k_p_indices = global_top_k_p_indices - vocab_start_index
+    local_top_k_p_indices = torch.where(
+        mask_top_k_p_indices, local_top_k_p_indices, torch.tensor(-1, dtype=torch.long)
+    )  # [B,S,k] - the global top_k_p indices are localized to the rank, or -1 if they are not in this rank's segment
+
+    full_mask = torch.full_like(logits, -float("Inf"))
+    batch_indices, sequence_indices = torch.where(
+        mask_top_k_p_indices.any(dim=-1)
+    )  # collect all b,s indices where there is a valid index in [b,s,:]
+    local_vocab_indices = local_top_k_p_indices[
+        batch_indices, sequence_indices
+    ]  # collect the v indices per each [b,s]. should be up to k valid indices (not valid is -1)
+    # is there a better way to do it ?
```

Comment:
Not sure -- I gave it a bit of thought but it's a tricky one. I think you did a good job at finding a way to do it with torch operations while minimizing the amount of tensor assignments (compared to the NeMo implementation that has a `for` loop and manually erases the removed logits)

The drawback is the large memory footprint of the `full_mask` tensor. However, I don't see how we can avoid this since we need to be able to backprop through these operations => we need a way to remember which logits were erased, however we do it, and this will be costly. Maybe one potential memory optimization would be instead to compute a boolean mask (4x less memory-hungry than a float32 tensor) and modify the logits in-place. Though with backprop shenanigans I wouldn't necessarily trust that it'd be more efficient without actually profiling the memory usage in practice.

(NB: the logic looks correct to me but I didn't actually run it with a TP model to confirm it works)

Reply:
OK, so should I keep it as is ? 
the other option, if I understand you correctly is to define binary full_mask and compute it such that 'True' will indicate that the value should be masked (i.e. -Inf) and if `False` should be remain as is and then apply this mask to logits, right ?